### PR TITLE
common : only load backends when required

### PR DESCRIPTION
**Pull request by angt**
## Overview

Only load backends when required.

## Additional information

This fixe the following issues: #20186, #21708 and maybe others.

# Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES (kimi): I've asked the list of functions to modify for a minimal patch

### CODE REVIEW

File: ggml/src/ggml-backend-reg.cpp [hunk 1]
```diff
@@ -541,6 +541,11 @@ static ggml_backend_reg_t ggml_backend_load_best(const char * name, bool silent,
 }
 
 void ggml_backend_load_all() {
+    static bool loaded = false;
```
Comment (danbev) [inline ggml/src/ggml-backend-reg.cpp]: Instead of modifying ggml could we perhaps introduce a helper function in arg.cpp that does this?

File: common/arg.cpp [hunk 2]
```diff
@@ -1002,9 +1005,6 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
 
     params.use_color = tty_can_use_colors();
 
-    // load dynamic backends
-    ggml_backend_load_all();
```
Comment (danbev) [inline common/arg.cpp:1020]: I think we need this in case none of the other options are used?

File: ggml/src/ggml-backend-reg.cpp (hunk 1 above)
Comment (angt) [reply ggml/src/ggml-backend-reg.cpp]: I opted for the minimal change but honestly this function should be able to be called many times without the static variable (I was a bit lazy) and I'm not sure adding a helper to hide that would be any better. I'm going to make a quick pre-PR.

File: common/arg.cpp (hunk 2 above)
Comment (angt) [reply common/arg.cpp:1020]: I think the current logic covers all scenarios do you have a specific edge case in mind ?

**Comment by ggerganov** [comment]
The `test-thread-safety` is failing

**Comment by ggerganov** [comment]
Hm, I don't see where we would call `ggml_backend_load_all()` for `llama-cli`, `llama-server`, etc. Building with `-DGGML_BACKEND_DL=ON` makes these binaries not load any backends.

**Comment by angt** [comment]
> Hm, I don't see where we would call `ggml_backend_load_all()` for `llama-cli`, `llama-server`, etc. Building with `-DGGML_BACKEND_DL=ON` makes these binaries not load any backends.

Damn, checking. I don’t use AI often (for llama.cpp!). Kimi is this bad ?  🤣

**Comment by angt** [comment]
The last call was missing...
Edit: So Kimi is fine, it was my fault 🙄

**Comment by angt** [comment]
@ggerganov maybe I can take this PR as an opportunity to remove `ggml_backend_load_all` from `common_params_parser_init` and call it explicitly where it’s needed, instead of hiding it there?

**Comment by ggerganov** [comment]
> @ggerganov maybe I can take this PR as an opportunity to remove `ggml_backend_load_all` from `common_params_parser_init` and call it explicitly where it’s needed, instead of hiding it there?

Yes, let's do that.

**Comment by angt** [comment]
I finally have updated `llama_backend_init()` to call `ggml_backend_load_all()`. The name was a perfect fit, so I decided to try this, but I’m not yet 100% convinced too, and maybe we should just call `ggml_backend_load_all()` everywhere...

**Review by ggerganov — APPROVED** [review]

**Review by pwilkin — APPROVED** [review]

**Review by danbev — APPROVED** [review]

**Comment by rgerganov** [comment]
This patch broke RPC (issue #22721) and introduced obscure regressions (issue #22748). Having multiple calls to `ggml_backend_load_all()` is not how loading dynamic backends is supposed to work and breaks many implicit assumptions.

I suggest to revert this patch.

**Comment by angt** [comment]
I think it’s just because of some missing calls it’s hard to track all of them

**Comment by angt** [comment]
#22752 should fix both issues (and others)
